File: py/safe_ci_dag_runner/ambient.py
"""Best-effort ambient host-load measurements.

The module reads Linux ``/proc`` counters, captures pressure and co-tenant activity,
and classifies a run as quiet, moderate, or busy. Missing or malformed measurements
produce explicit optional values or documented defaults; the readers never modify
host state.
"""

from __future__ import annotations

import os
from collections.abc import Collection
from dataclasses import dataclass
from pathlib import Path
from typing import Final, Literal

__all__ = [
    "PsiReading",
    "AmbientSnapshot",
    "AmbientBucket",
    "PROC_STAT_PATH",
    "PROC_CPU_PRESSURE_PATH",
    "PROC_MEMORY_PRESSURE_PATH",
    "PROC_IO_PRESSURE_PATH",
    "read_loadavg",
    "host_busy_jiffies",
    "read_pressure",
    "count_external_build_processes",
    "capture_ambient_snapshot",
    "attribute_external_cores",
    "ambient_bucket",
]

# Canonical /proc locations, named so callers/tests can point at fixtures.
PROC_STAT_PATH: Final = Path("/proc/stat")
PROC_CPU_PRESSURE_PATH: Final = Path("/proc/pressure/cpu")
PROC_MEMORY_PRESSURE_PATH: Final = Path("/proc/pressure/memory")
PROC_IO_PRESSURE_PATH: Final = Path("/proc/pressure/io")

# ``ambient_bucket`` cut-offs. PRESERVE THESE EXACTLY -- a sibling implementation in
# another language must produce the identical bucket for the identical inputs.
_BUSY_EXTERNAL_CORES: Final = 2.0
_BUSY_PSI_AVG10: Final = 20.0
_BUSY_CO_TENANTS: Final = 8
_QUIET_EXTERNAL_CORES: Final = 0.5
_QUIET_PSI_AVG10: Final = 5.0
_QUIET_CO_TENANTS: Final = 2

# Fallback USER_HZ when ``os.sysconf("SC_CLK_TCK")`` is unavailable, matching the
# DeepScry original's ``os.sysconf(...) or 100``.
_DEFAULT_CLK_TCK: Final = 100

#: The three-level ambient-load verdict, by string value (the value is the parity key).
AmbientBucket = Literal["quiet", "moderate", "busy"]


@dataclass(frozen=True)
class PsiReading:
    """One Pressure-Stall-Information (PSI) ``some`` line: the ``avg10`` / ``avg60``
    fractions of time at least one task was stalled on a resource.

    ``None`` rather than an empty reading marks an absent or unparseable
    ``/proc/pressure/*`` file, allowing callers to distinguish zero pressure from an
    unavailable kernel metric.
    """

    avg10: float
    avg60: float


@dataclass(frozen=True)
class AmbientSnapshot:
    """One instant of host-wide load used to attribute contention around a step."""

    #: System-wide non-idle CPU jiffies (``/proc/stat`` ``cpu`` line), or ``None`` when
    #: unreadable. Only meaningful as a delta between two snapshots.
    busy_jiffies: int | None
    #: 1- and 5-minute load averages (``os.getloadavg``).
    load1: float
    load5: float
    #: Host CPU / memory / IO pressure, each ``None`` when that PSI file is absent.
    cpu_psi: PsiReading | None
    memory_psi: PsiReading | None
    io_psi: PsiReading | None
    #: Count of matching build processes running OUTSIDE this run's cgroup scope.
    co_tenants: int


def read_loadavg() -> tuple[float, float, float]:
    """Return the 1/5/15-minute load averages via :func:`os.getloadavg`.

    Raises ``OSError`` on platforms that cannot report load averages.
    """
    load1, load5, load15 = os.getloadavg()
    return load1, load5, load15


def host_busy_jiffies(stat_path: Path = PROC_STAT_PATH) -> int | None:
    """Sum the system-wide non-idle CPU jiffies from ``/proc/stat``'s first ``cpu`` line.

    ``busy = total - idle - iowait`` (idle is field 3, iowait field 4 after the ``cpu``
    label). Returns ``None`` on any read or parse error rather than fabricating a count.
    """
    try:
        parts = stat_path.read_text().splitlines()[0].split()
        values = [int(value) for value in parts[1:]]
        return sum(values) - values[3] - (values[4] if len(values) > 4 else 0)
    except (OSError, ValueError, IndexError):
        return None


def read_pressure(path: Path) -> PsiReading | None:
    """Parse the ``some`` line of a PSI file (``/proc/pressure/{cpu,memory,io}``).

    Returns a :class:`PsiReading` with the ``avg10`` / ``avg60`` fractions, or ``None``
    when the file is missing, has no ``some`` line, or is malformed.
    """
    try:
        some = next(line for line in path.read_text().splitlines() if line.startswith("some "))
        values = dict(item.split("=", 1) for item in some.split()[1:])
        return PsiReading(avg10=float(values["avg10"]), avg60=float(values["avg60"]))
    except (OSError, StopIteration, ValueError, KeyError):
        return None


def count_external_build_processes(
    build_process_names: Collection[str],
    *,
    scope_marker: str | None = None,
) -> int:
    """Count named processes running outside this run's cgroup scope.

    * ``build_process_names`` -- exact ``/proc/<pid>/comm`` values to count. Kernel
      truncates ``comm`` to 15 bytes, so pass already-truncated names for long binaries.
    * ``scope_marker`` -- a substring identifying THIS run's cgroup. A process is counted
      only when the marker does NOT appear in its ``/proc/<pid>/cgroup`` (i.e. it is a
      genuine co-tenant, not one of our own children). ``None`` / empty counts every
      matching process, ours included.

    Per-process read errors (races on exiting pids, permission) are skipped silently by
    design: they are transient and cannot be attributed. The aggregate count is always
    returned.
    """
    names = set(build_process_names)
    if not names:
        # No build-process names to match => nothing to count. Return early so a caller that only
        # wants the load/PSI parts of a snapshot (e.g. the generic runner, which has no project
        # build-process names) does not pay for a full /proc scan that could only ever return 0.
        return 0
    marker = scope_marker or ""
    count = 0
    for entry in Path("/proc").iterdir():
        if not entry.name.isdigit():
            continue
        try:
            comm = (entry / "comm").read_text().strip()
            cgroup = (entry / "cgroup").read_text()
        except OSError:
            continue
        if comm in names and (not marker or marker not in cgroup):
            count += 1
    return count


def capture_ambient_snapshot(
    build_process_names: Collection[str],
    *,
    scope_marker: str | None = None,
) -> AmbientSnapshot:
    """Take one :class:`AmbientSnapshot` of current host load.

    Reads load averages, busy jiffies, the three host PSI files, and the co-tenant build count. The
    ``build_process_names`` / ``scope_marker`` arguments are forwarded verbatim to
    :func:`count_external_build_processes`.
    """
    load1, load5, _ = read_loadavg()
    return AmbientSnapshot(
        busy_jiffies=host_busy_jiffies(),
        load1=load1,
        load5=load5,
        cpu_psi=read_pressure(PROC_CPU_PRESSURE_PATH),
        memory_psi=read_pressure(PROC_MEMORY_PRESSURE_PATH),
        io_psi=read_pressure(PROC_IO_PRESSURE_PATH),
        co_tenants=count_external_build_processes(build_process_names, scope_marker=scope_marker),
    )


def _clk_tck() -> int:
    """USER_HZ (jiffies per second), falling back to 100 when ``sysconf`` cannot report."""
    try:
        return os.sysconf("SC_CLK_TCK") or _DEFAULT_CLK_TCK
    except (ValueError, OSError):
        return _DEFAULT_CLK_TCK


def attribute_external_cores(
    *,
    busy_jiffies_start: int | None,
    busy_jiffies_end: int | None,
    own_cpu_usec: int,
    elapsed_s: float,
) -> float:
    """Estimate how many CPU cores OTHER tenants burned during a step window.

    * ``host_busy_s`` -- system-wide non-idle CPU seconds over the window, from the
      busy-jiffies delta divided by USER_HZ. ``0`` if either endpoint is missing.
    * ``external_cpu_s = max(0, host_busy_s - own_cpu_usec/1e6)`` -- host CPU not
      attributable to this run's own cgroup.
    * ``external_cores = external_cpu_s / elapsed_s`` -- that CPU expressed as cores.

    Returns ``0.0`` for a non-positive ``elapsed_s`` because there is no window to divide by.
    """
    if busy_jiffies_start is not None and busy_jiffies_end is not None:
        host_busy_s = (busy_jiffies_end - busy_jiffies_start) / _clk_tck()
    else:
        host_busy_s = 0.0
    external_cpu_s = max(0.0, host_busy_s - own_cpu_usec / 1_000_000)
    if elapsed_s <= 0.0:
        return 0.0
    return external_cpu_s / elapsed_s


def ambient_bucket(external_cores: float, snapshot: AmbientSnapshot) -> AmbientBucket:
    """Classify host load as ``"quiet"`` / ``"moderate"`` / ``"busy"``.

    * ``busy``  -- external cores > 2.0, OR any host PSI ``avg10`` >= 20, OR co-tenants >= 8.
    * ``quiet`` -- external cores < 0.5, AND max PSI ``avg10`` < 5, AND co-tenants <= 2.
    * ``moderate`` -- everything in between.

    A missing PSI reading contributes ``avg10`` 0.0.
    """
    max_avg10 = max(
        (psi.avg10 for psi in (snapshot.cpu_psi, snapshot.memory_psi, snapshot.io_psi) if psi is not None),
        default=0.0,
    )
    co_tenants = snapshot.co_tenants
    if (
        external_cores > _BUSY_EXTERNAL_CORES
        or max_avg10 >= _BUSY_PSI_AVG10
        or co_tenants >= _BUSY_CO_TENANTS
    ):
        return "busy"
    if (
        external_cores < _QUIET_EXTERNAL_CORES
        and max_avg10 < _QUIET_PSI_AVG10
        and co_tenants <= _QUIET_CO_TENANTS
    ):
        return "quiet"
    return "moderate"

File: rs/safe-ci-dag-runner/src/ambient.rs
//! Ambient host-load capture: pure `/proc` readers plus a quiet/moderate/busy verdict.
//!
//! Rust port of `ambient.py`. Everything here reads only `/proc` and `os` counters; nothing
//! writes, forks, or touches cgroupfs. The `ambient_bucket` cut-offs are a cross-language parity
//! contract and are preserved verbatim in the named constants below.

use std::fs;
use std::path::{Path, PathBuf};

/// `/proc/stat`.
pub const PROC_STAT_PATH: &str = "/proc/stat";
/// `/proc/pressure/cpu`.
pub const PROC_CPU_PRESSURE_PATH: &str = "/proc/pressure/cpu";
/// `/proc/pressure/memory`.
pub const PROC_MEMORY_PRESSURE_PATH: &str = "/proc/pressure/memory";
/// `/proc/pressure/io`.
pub const PROC_IO_PRESSURE_PATH: &str = "/proc/pressure/io";

// ambient_bucket cut-offs. PRESERVE THESE EXACTLY (cross-language parity with ambient.py).
const BUSY_EXTERNAL_CORES: f64 = 2.0;
const BUSY_PSI_AVG10: f64 = 20.0;
const BUSY_CO_TENANTS: i64 = 8;
const QUIET_EXTERNAL_CORES: f64 = 0.5;
const QUIET_PSI_AVG10: f64 = 5.0;
const QUIET_CO_TENANTS: i64 = 2;
/// Fallback USER_HZ when `sysconf(SC_CLK_TCK)` is unavailable (matches Python's `or 100`).
const DEFAULT_CLK_TCK: i64 = 100;

/// The three-level ambient-load verdict (the string value is the parity key).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AmbientBucket {
    Quiet,
    Moderate,
    Busy,
}

impl AmbientBucket {
    /// Canonical string form matching Python's `AmbientBucket` literals.
    pub fn value(self) -> &'static str {
        match self {
            AmbientBucket::Quiet => "quiet",
            AmbientBucket::Moderate => "moderate",
            AmbientBucket::Busy => "busy",
        }
    }
}

/// One Pressure-Stall-Information `some` line (`avg10` / `avg60` fractions).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PsiReading {
    pub avg10: f64,
    pub avg60: f64,
}

/// One instant of host-wide load (the typed record `capture_ambient_snapshot` returns).
#[derive(Debug, Clone)]
pub struct AmbientSnapshot {
    /// System-wide non-idle CPU jiffies (`/proc/stat` `cpu` line), or `None` when unreadable.
    pub busy_jiffies: Option<i64>,
    pub load1: f64,
    pub load5: f64,
    pub cpu_psi: Option<PsiReading>,
    pub memory_psi: Option<PsiReading>,
    pub io_psi: Option<PsiReading>,
    /// Count of matching build processes running OUTSIDE this run's cgroup scope.
    pub co_tenants: i64,
}

/// USER_HZ (jiffies per second), falling back to 100 when it cannot be read.
fn clk_tck() -> i64 {
    // The `/proc/self/stat`-based tick is not exposed by std; read it via `getconf` when present,
    // otherwise use the conventional 100 Hz (matching Python's `os.sysconf(...) or 100` fallback
    // on hosts where sysconf is unavailable).
    if let Ok(out) = std::process::Command::new("getconf")
        .arg("CLK_TCK")
        .output()
    {
        if out.status.success() {
            if let Ok(n) = String::from_utf8_lossy(&out.stdout).trim().parse::<i64>() {
                if n > 0 {
                    return n;
                }
            }
        }
    }
    DEFAULT_CLK_TCK
}

/// 1/5/15-minute load averages from `/proc/loadavg` (Linux `os.getloadavg` analogue).
pub fn read_loadavg() -> (f64, f64, f64) {
    if let Ok(text) = fs::read_to_string("/proc/loadavg") {
        let mut parts = text.split_whitespace();
        let l1 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0.0);
        let l5 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0.0);
        let l15 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0.0);
        return (l1, l5, l15);
    }
    (0.0, 0.0, 0.0)
}

/// System-wide non-idle CPU jiffies from `/proc/stat`'s first `cpu` line.
///
/// `busy = total - idle - iowait` (idle is field 3, iowait field 4 after `cpu`). `None` on any
/// read/parse error, matching `host_busy_jiffies` in ambient.py.
pub fn host_busy_jiffies(stat_path: &Path) -> Option<i64> {
    let text = fs::read_to_string(stat_path).ok()?;
    let first = text.lines().next()?;
    let values: Vec<i64> = first
        .split_whitespace()
        .skip(1)
        .map(|v| v.parse::<i64>())
        .collect::<Result<_, _>>()
        .ok()?;
    if values.len() < 4 {
        return None;
    }
    let idle = values[3];
    let iowait = if values.len() > 4 { values[4] } else { 0 };
    Some(values.iter().sum::<i64>() - idle - iowait)
}

/// Parse the `some` line of a PSI file, or `None` when absent/malformed.
pub fn read_pressure(path: &Path) -> Option<PsiReading> {
    let text = fs::read_to_string(path).ok()?;
    let some = text.lines().find(|l| l.starts_with("some "))?;
    let mut avg10: Option<f64> = None;
    let mut avg60: Option<f64> = None;
    for item in some.split_whitespace().skip(1) {
        if let Some((k, v)) = item.split_once('=') {
            match k {
                "avg10" => avg10 = v.parse().ok(),
                "avg60" => avg60 = v.parse().ok(),
                _ => {}
            }
        }
    }
    Some(PsiReading {
        avg10: avg10?,
        avg60: avg60?,
    })
}

/// Count build processes whose `comm` is in `build_process_names` and that run OUTSIDE this
/// run's cgroup scope (`scope_marker` absent from `/proc/<pid>/cgroup`). Generic port of
/// `count_external_build_processes`.
pub fn count_external_build_processes(
    build_process_names: &[&str],
    scope_marker: Option<&str>,
) -> i64 {
    let names: std::collections::HashSet<&str> = build_process_names.iter().copied().collect();
    if names.is_empty() {
        // No build-process names to match => nothing to count. Return early so a caller that only
        // wants the load/PSI parts of a snapshot (the generic runner, which has no project
        // build-process names) does not pay for a full /proc scan that could only ever return 0.
        return 0;
    }
    let marker = scope_marker.unwrap_or("");
    let entries = match fs::read_dir("/proc") {
        Ok(e) => e,
        Err(_) => return 0,
    };
    let mut count = 0;
    for entry in entries.flatten() {
        let name = entry.file_name();
        let name = match name.to_str() {
            Some(n) if n.chars().all(|c| c.is_ascii_digit()) => n.to_string(),
            _ => continue,
        };
        let dir = PathBuf::from("/proc").join(&name);
        let comm = match fs::read_to_string(dir.join("comm")) {
            Ok(c) => c.trim().to_string(),
            Err(_) => continue,
        };
        let cgroup = match fs::read_to_string(dir.join("cgroup")) {
            Ok(c) => c,
            Err(_) => continue,
        };
        if names.contains(comm.as_str()) && (marker.is_empty() || !cgroup.contains(marker)) {
            count += 1;
        }
    }
    count
}

/// Take one snapshot of current host load (typed port of `capture_ambient_snapshot`).
pub fn capture_ambient_snapshot(
    build_process_names: &[&str],
    scope_marker: Option<&str>,
) -> AmbientSnapshot {
    let (load1, load5, _) = read_loadavg();
    AmbientSnapshot {
        busy_jiffies: host_busy_jiffies(Path::new(PROC_STAT_PATH)),
        load1,
        load5,
        cpu_psi: read_pressure(Path::new(PROC_CPU_PRESSURE_PATH)),
        memory_psi: read_pressure(Path::new(PROC_MEMORY_PRESSURE_PATH)),
        io_psi: read_pressure(Path::new(PROC_IO_PRESSURE_PATH)),
        co_tenants: count_external_build_processes(build_process_names, scope_marker),
    }
}

/// Estimate how many CPU cores OTHER tenants burned during a step window (port of the
/// arithmetic in `attribute_external_cores`). Returns `0.0` for a non-positive `elapsed_s`.
pub fn attribute_external_cores(
    busy_jiffies_start: Option<i64>,
    busy_jiffies_end: Option<i64>,
    own_cpu_usec: i64,
    elapsed_s: f64,
) -> f64 {
    let host_busy_s = match (busy_jiffies_start, busy_jiffies_end) {
        (Some(a), Some(b)) => (b - a) as f64 / clk_tck() as f64,
        _ => 0.0,
    };
    let external_cpu_s = (host_busy_s - own_cpu_usec as f64 / 1_000_000.0).max(0.0);
    if elapsed_s <= 0.0 {
        return 0.0;
    }
    external_cpu_s / elapsed_s
}

/// Classify host load as quiet / moderate / busy. Thresholds preserved EXACTLY (parity with
/// `ambient_bucket` in ambient.py). A missing PSI reading contributes `avg10` 0.0.
pub fn ambient_bucket(external_cores: f64, snapshot: &AmbientSnapshot) -> AmbientBucket {
    let max_avg10 = [snapshot.cpu_psi, snapshot.memory_psi, snapshot.io_psi]
        .into_iter()
        .flatten()
        .map(|p| p.avg10)
        .fold(0.0_f64, f64::max);
    let co = snapshot.co_tenants;
    if external_cores > BUSY_EXTERNAL_CORES || max_avg10 >= BUSY_PSI_AVG10 || co >= BUSY_CO_TENANTS
    {
        return AmbientBucket::Busy;
    }
    if external_cores < QUIET_EXTERNAL_CORES
        && max_avg10 < QUIET_PSI_AVG10
        && co <= QUIET_CO_TENANTS
    {
        return AmbientBucket::Quiet;
    }
    AmbientBucket::Moderate
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(cpu_avg10: f64, co: i64) -> AmbientSnapshot {
        AmbientSnapshot {
            busy_jiffies: Some(0),
            load1: 0.0,
            load5: 0.0,
            cpu_psi: Some(PsiReading {
                avg10: cpu_avg10,
                avg60: 0.0,
            }),
            memory_psi: None,
            io_psi: None,
            co_tenants: co,
        }
    }

    #[test]
    fn bucket_thresholds_match_reference() {
        // busy: external cores > 2.0
        assert_eq!(ambient_bucket(2.5, &snap(0.0, 0)), AmbientBucket::Busy);
        // busy: any PSI avg10 >= 20
        assert_eq!(ambient_bucket(0.0, &snap(20.0, 0)), AmbientBucket::Busy);
        // busy: co-tenants >= 8
        assert_eq!(ambient_bucket(0.0, &snap(0.0, 8)), AmbientBucket::Busy);
        // quiet: all below the quiet cut-offs
        assert_eq!(ambient_bucket(0.4, &snap(4.9, 2)), AmbientBucket::Quiet);
        // moderate: in between
        assert_eq!(ambient_bucket(1.0, &snap(10.0, 3)), AmbientBucket::Moderate);
    }

    #[test]
    fn attribute_external_cores_guards_zero_window() {
        assert_eq!(attribute_external_cores(Some(0), Some(100), 0, 0.0), 0.0);
        // With a positive window: (100 jiffies / clk) - 0 own, over 1s.
        let cores = attribute_external_cores(Some(0), Some(clk_tck()), 0, 1.0);
        assert!((cores - 1.0).abs() < 1e-9);
    }
}
